src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,147 @@
+        // isolates the command word
+        String[] words = input.split(" ");
+        String command = words[0];
+
+        try {
+            switch (command) {
+            case "bye": // only applicable to GUI Duke
+                storage.save(tasks.getTaskList());
+                return ui.exit();
+            case "list":
+                return tasks.displayList();
+            case "done":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // extracts index of task to mark as done
+                String doneTaskIndex = words[words.length - 1];
+                return tasks.markDone(doneTaskIndex);
+            case "deadline":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                try {
+                    // excludes command "deadline " from the string
+                    return tasks.addDeadline(input.substring(9));
+                } catch (InvalidDateTimeException | MessageEmptyException | IncorrectFormatException e) {
+                    return e.getMessage();
+                }
+            case "todo":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // excludes command "todo" from the string
+                return tasks.addTodo(input.substring(5));
+            case "event":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // excludes command "event" from the string
+                return tasks.addEvent(input.substring(6));
+            case "delete":
+                if (words.length == 1) {
+                    // throws an error if there is no message input after the command word
+                    throw new MessageEmptyException();
+                }
+                // extracts index of task to delete
+                String deleteTaskIndex = words[words.length - 1];
+                return tasks.deleteTask(deleteTaskIndex);
+            case "find":
+                ArrayList<Task> matchedTasks = findCommand(words);
+                return Ui.printList(matchedTasks);
+            case "": // empty user input
+                throw new EmptyCommandException();
+            default: // all other inputs that are not supported
+                throw new InvalidCommandException();
+            }
+        } catch (DukeException | IOException e) {
+            return e.getMessage();
+        }
```

Comment:
I think a switch-case flow might get messy if we have to continue adding more commands.
Maybe consider using a `Hashtable<String, Function>`? 

Alternatively, maybe handle error-checking within the methods invoked (eg. `tasks.deleteTask()`) instead of outside? I think it makes the code more readable because it trims the switch-case block and it complies to OOP's principle of encapsulation as well :)